src/handler/mod.rs

```diff
@@ -813,6 +828,37 @@ impl Handler {
                         // established. If so process them.
                         self.send_next_request::<P>(node_address).await;
                     } else {
+                        // Respond to PING request even if the ENR or NodeAddress don't match
```

Comment:
Here if we store just one one-time-session per peer, we would want to check first for one such session being already established before doing the decode

Reply:
> we would want to check first for one such session being already established

You mean we would want to move the second statement of below (`else if ... self.remove_one_time_session() ...`) to the top of the if statement?

https://github.com/sigp/discv5/blob/fce0fce685ad7aae4193124123996f85b7f30beb/src/handler/mod.rs#L522-L535

Reply:
yep, this is what I mean

Reply:
My slight concern about doing that is, the second statement (`if ... self.remove_one_time_session() ...`) is false in most cases. Establishing a one-time session is under limited conditions, so the result of the second statement should be false in most send_response() calls.

Do you have a good reason for moving the second statement to the first?

Reply:
My reasoning is that is not impossible to have a one-time session created for a peer, then a normal session created, and then hit this line. A bit of a race condition. Moving the check up would prevent this. It's a matter of correctness but I understand your pov as a matter of performance. What are your thoughts on this?

Reply:
I see your point. Your suggestion is right in terms of correctness. 

> a one-time session created for a peer, then a normal session created, and then hit this line

Yeah it looks a bit of a race condition but I think that if a normal session created, it's fine to use the normal session even if a one-time session (for the request) exists since the normal session is valid as well. The one-time session will be evicted over time without being used.

Reply:
@AgeManning do you have any opinion here?

Reply:
Not really. 
Its a single hash lookup that we are worried about. I think if we have a session, we shouldn't establish a one-time session. But there is a race condition where a session gets created after. So I think its a relatively rare case that we have both sessions. In this case, we probably want to prioritise the normal session over the one-time session. 

Seeing as we are probably running this a bit, I agree with @ackintosh in that it's nicer to avoid the hash lookup (i.e check and remove the one-time session) to handle this edge case. It's also no real loss because the one-time sessions only last 30 seconds (assuming they get pruned). 

But imo we are debating over a single hash lookup vs correctness, so i'm really on the fence. 

Maybe just leave it as @ackintosh has it, because we don't have to do another commit? 
I don't feel strongly at all, if either one of you do, happy to go that way.

Reply:
also of no strong opinion so merged :shrug: seems we are all happy so